// grid-assign-js/dist/grid/index.mjs
import { createCommonjsModule, unwrapExports, commonjsGlobal } from '../../../../_virtual/_commonjsHelpers.mjs';
import { __moduleExports as lapJv } from '../lap-jv/index.mjs';

var grid = createCommonjsModule(function (module, exports) {
var __importDefault = (commonjsGlobal && commonjsGlobal.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.autoGridAssignment = exports.generateGrid = void 0;
const lap_jv_1 = __importDefault(lapJv);
function generateGrid({ width, height, numRows, numColumns, centerPoint = false, translate = [0, 0], }) {
    const gridHeight = height / numRows;
    const gridWidth = width / numColumns;
    const gridPoints = [];
    let centerTranslate = [0, 0];
    if (centerPoint) {
        centerTranslate[0] = gridWidth / 2;
        centerTranslate[1] = gridHeight / 2;
    }
    let yPosition = translate[0];
    for (let r = 0; r < numRows; r++) {
        let xPosition = translate[1];
        for (let c = 0; c < numColumns; c++) {
            const gridPoint = [
                xPosition + centerTranslate[0],
                yPosition + centerTranslate[1],
            ];
            gridPoints.push(gridPoint);
            xPosition += gridWidth;
        }
        yPosition += gridHeight;
    }
    return { gridHeight, gridWidth, gridPoints };
}
exports.generateGrid = generateGrid;
function autoGridAssignment({ points, screenWidth, screenHeight, screenX = 0, screenY = 0, numColumns, numRows, anchorMiddleGrid = false, }) {
    // compute the grid on top of the points
    const { gridWidth, gridHeight, gridPoints } = generateGrid({
        height: screenHeight,
        width: screenWidth,
        numColumns,
        numRows,
        translate: [screenX, screenY],
        centerPoint: anchorMiddleGrid,
    });
    // compute min max of the x and y directions of the points
    const pointExtent = computeMinMax({ points, accessor: (d) => d });
    const gridExtent = {
        x: [screenX, screenX + screenWidth],
        y: [screenY, screenY + screenHeight],
    };
    const convert = {
        pointToScreen: {
            x: scaleLinear({ domain: pointExtent.x, range: gridExtent.x }),
            y: scaleLinear({ domain: pointExtent.y, range: gridExtent.y }),
        },
        screenToPoint: {
            x: scaleLinear({ domain: gridExtent.x, range: pointExtent.x }),
            y: scaleLinear({ domain: gridExtent.y, range: pointExtent.y }),
        },
        apply2D: (point, converter) => [converter.x(point[0]), converter.y(point[1])],
    };
    const gridsInPointSpace = gridPoints.map((p) => convert.apply2D(p, convert.screenToPoint));
    // apply linear assignment to get optimal assignments
    const assignments = (0, lap_jv_1.default)({ points, assignTo: gridsInPointSpace });
    // convert back to the screen space grids and give assignments
    const screenSpaceAssignments = assignments.map((a, i) => {
        return {
            gridPoint: gridPoints[i],
            gridWidth,
            gridHeight,
            assignedPoint: points[a],
            assignedPointIndex: a,
        };
    });
    return { assignments: screenSpaceAssignments, pointConversions: convert };
}
exports.autoGridAssignment = autoGridAssignment;
/**
 * Given a point within the domain output a mapping to the range
 * *linearly*
 */
function scaleLinear({ domain = [0, 0], range = [0, 0] }) {
    const x1 = domain[0], y1 = range[0];
    const x2 = domain[1], y2 = range[1];
    const deltaX = x2 - x1;
    if (deltaX === 0) {
        throw Error("cannot divide by 0 in slope calculation");
    }
    const deltaY = y2 - y1;
    // y = mx + b
    const m = deltaX / deltaY;
    const b = y1 - m * x1;
    const y = (x) => m * x + b;
    return y;
}
function computeMinMax({ points, accessor }) {
    const xExtent = [Infinity, -Infinity];
    const yExtent = [Infinity, -Infinity];
    points.forEach((d) => {
        const point = accessor(d);
        const [x, y] = point;
        if (x > xExtent[1]) {
            xExtent[1] = x;
        }
        if (x < xExtent[0]) {
            xExtent[0] = x;
        }
        if (y > yExtent[1]) {
            yExtent[1] = y;
        }
        if (y < yExtent[0]) {
            yExtent[0] = y;
        }
    });
    return { x: xExtent, y: yExtent };
}

});

unwrapExports(grid);
grid.autoGridAssignment;
grid.generateGrid;

export { grid as __moduleExports };
